### PR TITLE
Canvas Components for OffscreenCanvas

### PR DESCRIPTION
**Pull request by maharsh312**
---

- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [ ] These changes fix #___ (GitHub issue number if applicable)

- [X] There are tests for these changes OR
- [ ] These changes do not require tests because ___

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/23381)

### CODE REVIEW

File: components/script/dom/canvasrenderingcontext2d.rs [hunk 1]
```diff
@@ -119,15 +114,32 @@ impl CanvasContextState {
...
     #[ignore_malloc_size_of = "Defined in ipc-channel"]
     ipc_renderer: IpcSender<CanvasMsg>,
     canvas_id: CanvasId,
+    state: DomRefCell<CanvasContextState>,
+    origin_clean: Cell<bool>,
+    canvas: Option<&HTMLCanvasElement>,
+    #[ignore_malloc_size_of = "Arc"]
+    image_cache: Arc<dyn ImageCache>,
+    base_url: ServoUrl,
+    global: &GlobalScope,
```
Comment (jdm) [inline components/script/dom/canvasrenderingcontext2d.rs]: Remove `global` and have callers pass in a `&GlobalScope` argument when necessary.

File: components/script/dom/canvasrenderingcontext2d.rs [hunk 2]
```diff
@@ -119,15 +114,32 @@ impl CanvasContextState {
     }
 }
 
+#[must_root]
 #[derive(JSTraceable, MallocSizeOf)]
 pub struct CanvasState {
     #[ignore_malloc_size_of = "Defined in ipc-channel"]
     ipc_renderer: IpcSender<CanvasMsg>,
     canvas_id: CanvasId,
+    state: DomRefCell<CanvasContextState>,
```
Comment (jdm) [inline components/script/dom/canvasrenderingcontext2d.rs]: We no longer need DomRefCell or Cell in members of CanvasState, since CanvasState is stored in a DomRefCell. We can write methods that require `&mut self` instead of `&self` when we need to mutate members now.

File: components/script/dom/canvasrenderingcontext2d.rs [hunk 3]
```diff
@@ -119,15 +114,32 @@ impl CanvasContextState {
...
+    origin_clean: Cell<bool>,
+    canvas: Option<&HTMLCanvasElement>,
```
Comment (jdm) [inline components/script/dom/canvasrenderingcontext2d.rs]: Have callers pass in a `Option<&HTMLCanvasElement>` argument when necessary instead of storing this.

**Review by jdm — CHANGES_REQUESTED** [review]
Fixing up the members of CanvasState should avoid many errors currently reported by the Rust compiler. In general, we can't store any pointers to objects (`&T`), so we should remove them entirely or store types like `Dom<T>`.

**Comment by jdm** [comment]
A rebase against master will be required.

**Comment by jdm** [comment]
I successfully cherry-picked and rebased the changes.

**Comment by maharsh312** [comment]
> I successfully cherry-picked and rebased the changes.

Thank You Josh!

**Comment by jdm** [comment]
I filed #23436 for some cleanup that can happen, but the implementation looks sound to me. Thanks for completing this work!

**Comment by maharsh312** [comment]
> I filed #23436 for some cleanup that can happen, but the implementation looks sound to me. Thanks for completing this work!

Thank you for helping me throughout this process!

**Comment by CYBAI** [comment]
linux builders failed with

```
error: method is never used: `image_cache`
   --> components/script/dom/paintworkletglobalscope.rs:134:5
    |
134 |     pub fn image_cache(&self) -> Arc<dyn ImageCache> {
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |
    = note: `-D dead-code` implied by `-D warnings`
```

and Travis complains about tidy

```
$ ./mach test-tidy --no-progress --all
 0:08.35 INFO Diffing old and new manifests /home/travis/build/servo/servo/tests/wpt/mozilla/meta/MANIFEST.json
 0:08.48 WARNING Manifest /home/travis/build/servo/servo/tests/wpt/mozilla/meta/MANIFEST.json contains correct tests but file hashes changed.
 0:08.48 ERROR Manifest /home/travis/build/servo/servo/tests/wpt/mozilla/meta/MANIFEST.json is outdated, use |./mach update-manifest| to fix.
 0:08.48 INFO Diffing old and new manifests /home/travis/build/servo/servo/tests/wpt/metadata/MANIFEST.json
 0:13.72 INFO Diffing old and new manifests /home/travis/build/servo/servo/tests/wpt/webgl/meta/MANIFEST.json
```

**Comment by jdm** [comment]
This is triggering a known edge case in our CI and it's going to be tricky to address. Go ahead and email your professor; don't wait for this PR to finish merging.

**Comment by maharsh312** [comment]
> This is triggering a known edge case in our CI and it's going to be tricky to address. Go ahead and email your professor; don't wait for this PR to finish merging.

Okay Josh!
